rfcs/20200810-tf-internal-api.md

```diff
@@ -0,0 +1,119 @@
+| **RFC #**     | [278](https://github.com/tensorflow/community/pull/278)|
+| **Author(s)** | Qianli Zhu (scottzhu@google.com) |
+| **Sponsor**   | Martin Wicke (wicke@google.com), Alex Apassos (apassos@google.com)|
+| **Updated**   | 2020-08-10                                           |
+| **Intended audience**| tf-api-owners, tf-addons, keras-team, deepmind/sonnet|
+
+## Objective
+
+Adding a new "internal" API namespace in TF to host APIs for framework building/testing, etc. The API will have looser contracts compared to core TF API.
+
+## Motivation
+
+TensorFlow has a strong contract to ensure the stability of its API. The core 
+API, once added, can't be updated with backward incompatible change or removed. 
+TF API owners will pay extra attention to any new API proposal due to this 
+restriction, based on the cost we have to bear to maintain the API 
+compatibility. See more details about API review 
+[here](https://github.com/tensorflow/community/blob/master/governance/api-reviews.md).
+
+Keras team is trying to split its code into a [standalone Github repository](https://github.com/tensorflow/community/blob/master/rfcs/20200205-standalone-keras-repository.md). 
+One of the key actions before the repository split is that Keras code has to 
+rely on the public TF API only, to avoid any future breakages due to private 
+method/class change.
+
+Historically, Keras was an internal component of TF, and relies heavily on a lot
+of private TF utilities and functionalities that are not part of the public API. 
+Exposing those functions into public APIs will be very confusing to the end 
+user, since those functions are useful to framework-builders, but not typical 
+TensorFlow users. In addition, these utilities would be overly restricted by the 
+standard TensorFlow API contract, as we would like to change/remove them more 
+often than the existing core API.
+
+
+## Design Proposal
+
+Add a new namespace "tensorflow.internal" to host the framework building/testing 
+related APIs. <b>(See alternative naming in the sections below, the naming of 
+the namespace will be discussed in the design meeting.)</b>
+
+### Backward compatibility
+The contract for all the APIs under this namespace is looser compared to core TF
+API. It will remain backward compatible for at least 1 minor TF release. 
+
+This means for any API that is added in 2.x release, it will remain the same in 
+2.x+1. If we choose to remove the same API, we will mark it as deprecated in 
+2.x+1, and delete it at 2.x+2 release. TensorFlow is released every 3 months, 
+and this will give enough time for all the clients to change to the new 
+API/alternative.
+
+Any deprecation and backward incompatible API change will be explicitly called 
+out in TF release notes. This applies to both core API and "internal" API.
+
+### Acceptance Criteria
+The candidate of the "internal" API should:
+
+1. Does NOT fit for core TF API, otherwise it should be exposed as core TF API.
```

Comment:
Can you expand on point 1? It Is very ambiguous.

Reply:
The general guild line for TF API is https://github.com/tensorflow/community/blob/master/governance/api-reviews.md. API owner group will discuss individual API proposal in the weekly meeting.

Reply:
The general guild line for TF API is https://github.com/tensorflow/community/blob/master/governance/api-reviews.md. API owner group will discuss individual API proposal in the weekly meeting.